File: scripts/manage-kind-hosts.rb
#!/usr/bin/env ruby

require "ipaddr"
require "open3"
require "optparse"

BEGIN_MARKER = "# BEGIN k8s-playground managed kind hosts"
END_MARKER = "# END k8s-playground managed kind hosts"
HOSTNAMES = %w[
  app.k8s-playground.test
  argocd.k8s-playground.test
  grafana.k8s-playground.test
].freeze

options = {
  context: "kind-k8s-playground",
  hosts_file: "/etc/hosts"
}

parser = OptionParser.new do |opts|
  opts.banner = "Usage: #{File.basename($PROGRAM_NAME)} MODE [options]"
  opts.on("--context CONTEXT", "Kubernetes context") { |value| options[:context] = value }
  opts.on("--hosts-file PATH", "Hosts file to inspect") { |value| options[:hosts_file] = value }
  opts.on("--user-gateway-ip IP", "Override the discovered user gateway IP") { |value| options[:user_ip] = value }
  opts.on("--management-gateway-ip IP", "Override the discovered management gateway IP") { |value| options[:management_ip] = value }
end

parser.parse!
mode = ARGV.shift
unless %w[check render remove].include?(mode) && ARGV.empty?
  warn parser
  exit 2
end

def gateway_ip(context, namespace, service)
  command = [
    "kubectl", "--context", context, "-n", namespace,
    "get", "service/#{service}",
    "-o", "jsonpath={.status.loadBalancer.ingress[0].ip}"
  ]
  output, error, status = Open3.capture3(*command)
  abort "Could not read #{namespace}/#{service}: #{error.strip}" unless status.success?

  ip = output.strip
  abort "#{namespace}/#{service} does not have a LoadBalancer IP" if ip.empty?
  ip
end

def validate_ipv4!(value, label)
  address = IPAddr.new(value)
  abort "#{label} is not an IPv4 address: #{value}" unless address.ipv4?
rescue IPAddr::InvalidAddressError
  abort "#{label} is not a valid IP address: #{value}"
end

def locate_managed_block(lines)
  starts = lines.each_index.select { |index| lines[index].chomp == BEGIN_MARKER }
  ends = lines.each_index.select { |index| lines[index].chomp == END_MARKER }

  if starts.length > 1 || ends.length > 1 || starts.length != ends.length
    abort "Managed hosts markers are duplicated or unbalanced; refusing to modify the file"
  end
  return nil if starts.empty?
  abort "Managed hosts markers are reversed; refusing to modify the file" if starts.first > ends.first

  starts.first..ends.first
end

def conflicting_hostnames(lines, managed_range)
  lines.each_with_index.flat_map do |line, index|
    next [] if managed_range&.cover?(index)

    fields = line.split("#", 2).first.to_s.split
    next [] if fields.length < 2

    fields.drop(1) & HOSTNAMES
  end.uniq
end

content = File.binread(options[:hosts_file])
lines = content.lines
managed_range = locate_managed_block(lines)
conflicts = conflicting_hostnames(lines, managed_range)
unless conflicts.empty?
  abort "Managed hostnames exist outside the managed block: #{conflicts.join(", ")}"
end

if mode == "remove"
  if managed_range
    $stdout.write(lines[0...managed_range.begin].join)
    $stdout.write(lines[(managed_range.end + 1)..].to_a.join)
  else
    $stdout.write(content)
  end
  exit 0
end

user_ip = options[:user_ip] || gateway_ip(options[:context], "istio-system", "istio-ingressgateway")
management_ip = options[:management_ip] || gateway_ip(options[:context], "istio-system", "istio-managementgateway")
validate_ipv4!(user_ip, "User gateway IP")
validate_ipv4!(management_ip, "Management gateway IP")

expected_block = <<~HOSTS
  #{BEGIN_MARKER}
  # Kubernetes context: #{options[:context]}
  #{user_ip} app.k8s-playground.test
  #{management_ip} argocd.k8s-playground.test grafana.k8s-playground.test
  #{END_MARKER}
HOSTS

if mode == "check"
  abort "Managed hosts block is missing" unless managed_range

  actual_block = lines[managed_range].join
  abort "Managed hosts block is stale" unless actual_block == expected_block

  puts "Managed hosts block is current for #{options[:context]}."
  exit 0
end

if managed_range
  $stdout.write(lines[0...managed_range.begin].join)
  $stdout.write(expected_block)
  $stdout.write(lines[(managed_range.end + 1)..].to_a.join)
else
  $stdout.write(content)
  $stdout.write("\n") unless content.empty? || content.end_with?("\n")
  $stdout.write("\n") unless content.empty? || content.end_with?("\n\n")
  $stdout.write(expected_block)
end

File: scripts/manage-kind-hosts.py
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.12"
# dependencies = []
# ///

import argparse
import ipaddress
import subprocess
import sys
from pathlib import Path


BEGIN_MARKER = "# BEGIN k8s-playground managed kind hosts"
END_MARKER = "# END k8s-playground managed kind hosts"
HOSTNAMES = {
    "app.k8s-playground.test",
    "argocd.k8s-playground.test",
    "grafana.k8s-playground.test",
}


def fail(message: str) -> None:
    print(message, file=sys.stderr)
    raise SystemExit(1)


def gateway_ip(context: str, namespace: str, service: str) -> str:
    command = [
        "kubectl",
        "--context",
        context,
        "-n",
        namespace,
        "get",
        f"service/{service}",
        "-o",
        "jsonpath={.status.loadBalancer.ingress[0].ip}",
    ]
    result = subprocess.run(command, capture_output=True, text=True, check=False)
    if result.returncode != 0:
        fail(f"Could not read {namespace}/{service}: {result.stderr.strip()}")

    value = result.stdout.strip()
    if not value:
        fail(f"{namespace}/{service} does not have a LoadBalancer IP")
    return value


def validate_ipv4(value: str, label: str) -> None:
    try:
        address = ipaddress.ip_address(value)
    except ValueError:
        fail(f"{label} is not a valid IP address: {value}")
    if address.version != 4:
        fail(f"{label} is not an IPv4 address: {value}")


def line_without_ending(line: bytes) -> bytes:
    return line.rstrip(b"\r\n")


def locate_managed_block(lines: list[bytes]) -> tuple[int, int] | None:
    starts = [index for index, line in enumerate(lines) if line_without_ending(line) == BEGIN_MARKER.encode()]
    ends = [index for index, line in enumerate(lines) if line_without_ending(line) == END_MARKER.encode()]

    if len(starts) > 1 or len(ends) > 1 or len(starts) != len(ends):
        fail("Managed hosts markers are duplicated or unbalanced; refusing to modify the file")
    if not starts:
        return None
    if starts[0] > ends[0]:
        fail("Managed hosts markers are reversed; refusing to modify the file")
    return starts[0], ends[0]


def conflicting_hostnames(lines: list[bytes], managed_range: tuple[int, int] | None) -> list[str]:
    conflicts: list[str] = []
    for index, line in enumerate(lines):
        if managed_range and managed_range[0] <= index <= managed_range[1]:
            continue
        fields = line.split(b"#", 1)[0].split()
        if len(fields) < 2:
            continue
        for field in fields[1:]:
            hostname = field.decode(errors="surrogateescape")
            if hostname in HOSTNAMES and hostname not in conflicts:
                conflicts.append(hostname)
    return conflicts


def expected_block(context: str, user_ip: str, management_ip: str) -> bytes:
    return (
        f"{BEGIN_MARKER}\n"
        f"# Kubernetes context: {context}\n"
        f"{user_ip} app.k8s-playground.test\n"
        f"{management_ip} argocd.k8s-playground.test grafana.k8s-playground.test\n"
        f"{END_MARKER}\n"
    ).encode()


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(usage="%(prog)s MODE [options]")
    parser.add_argument("mode", choices=("check", "render", "remove"))
    parser.add_argument("--context", default="kind-k8s-playground", help="Kubernetes context")
    parser.add_argument("--hosts-file", default="/etc/hosts", help="Hosts file to inspect")
    parser.add_argument("--user-gateway-ip", help="Override the discovered user gateway IP")
    parser.add_argument("--management-gateway-ip", help="Override the discovered management gateway IP")
    return parser.parse_args()


def main() -> None:
    args = parse_args()
    content = Path(args.hosts_file).read_bytes()
    lines = content.splitlines(keepends=True)
    managed_range = locate_managed_block(lines)
    conflicts = conflicting_hostnames(lines, managed_range)
    if conflicts:
        fail(f"Managed hostnames exist outside the managed block: {', '.join(conflicts)}")

    if args.mode == "remove":
        if managed_range:
            start, end = managed_range
            output = b"".join(lines[:start] + lines[end + 1 :])
        else:
            output = content
        sys.stdout.buffer.write(output)
        return

    user_ip = args.user_gateway_ip or gateway_ip(args.context, "istio-system", "istio-ingressgateway")
    management_ip = args.management_gateway_ip or gateway_ip(
        args.context, "istio-system", "istio-managementgateway"
    )
    validate_ipv4(user_ip, "User gateway IP")
    validate_ipv4(management_ip, "Management gateway IP")
    block = expected_block(args.context, user_ip, management_ip)

    if args.mode == "check":
        if not managed_range:
            fail("Managed hosts block is missing")
        start, end = managed_range
        if b"".join(lines[start : end + 1]) != block:
            fail("Managed hosts block is stale")
        print(f"Managed hosts block is current for {args.context}.")
        return

    if managed_range:
        start, end = managed_range
        output = b"".join(lines[:start]) + block + b"".join(lines[end + 1 :])
    else:
        separator = b""
        if content and not content.endswith(b"\n"):
            separator += b"\n"
        if content and not content.endswith(b"\n\n"):
            separator += b"\n"
        output = content + separator + block
    sys.stdout.buffer.write(output)


if __name__ == "__main__":
    main()
